mixer/pkg/cache/ttlCache.go

```diff
@@ -0,0 +1,144 @@
+//
+// Cache eviction is done on a periodic basis. Individual cache entries are evicted
+// after their expiration time has passed. The periodic nature of eviction means that
+// cache entries tend to survive around (expirationTime + (evictionInterval / 2))
+//
+// defaultExpiration specifies the default minimum amount of time a cached
+// entry remains in the cache before eviction. This value is used with the
+// Set function. Explicit per-entry expiration times can be set with the
+// SetWithExpiration function instead.
+//
+// evictionInterval specifies the frequency at which eviction activities take
+// place. This should be >= to 1 second.
+//
+// Since TTL caches only evict data based on the passage of time, it's possible to
+// use up all available memory by continuing to add entries to the cache with a
+// long enough expiration time. Don't do that.
+func NewTTLCache(defaultExpiration time.Duration, evictionInterval time.Duration) Cache {
+	c := &ttlCache{
+		defaultExpiration: defaultExpiration,
+	}
+
+	c.baseTime = time.Now().UTC().UnixNano()
+
+	if evictionInterval > 0 {
+		c.evictionTicker = time.NewTicker(evictionInterval)
+		go c.evicter()
+	}
+
+	return c
+}
+
+func (c ttlCache) Close() {
+	if c.evictionTicker != nil {
+		c.evictionTicker.Stop()
+	}
+}
+
+func (c *ttlCache) evicter() {
+	// Wake up once in a while and evict stale items
+	for now := range c.evictionTicker.C {
+		c.evictExpired(now)
+	}
+}
+
+func (c *ttlCache) evictExpired(t time.Time) {
+	// We snapshot a base time here such that the time doesn't need to be
+	// sampled in the Set call as calling time.Now() is relatively expensive.
+	// Doing it here provides enough precision for our needs and tends to have
+	// much lower call frequency.
+	n := t.UTC().UnixNano()
+	atomic.StoreInt64(&c.baseTime, n)
+
+	var count uint64
+
+	c.entries.Range(func(key interface{}, value interface{}) bool {
+		e := value.(*entry)
+		if e.expiration <= n {
+			c.entries.Delete(key)
+			count++
+		}
+		return true
+	})
+
+	atomic.AddUint64(&c.stats.Writes, count)
+}
+
+func (c *ttlCache) Set(key interface{}, value interface{}) {
+	c.SetWithExpiration(key, value, c.defaultExpiration)
+}
+
+func (c *ttlCache) SetWithExpiration(key interface{}, value interface{}, expiration time.Duration) {
+	e := &entry{
+		value:      value,
+		expiration: atomic.LoadInt64(&c.baseTime) + expiration.Nanoseconds(),
+	}
+
+	c.entries.Store(key, e)
+	atomic.AddUint64(&c.stats.Writes, 1)
+}
+
+func (c *ttlCache) Get(key interface{}) (interface{}, bool) {
+	e, ok := c.entries.Load(key)
+	if !ok {
+		atomic.AddUint64(&c.stats.Misses, 1)
+		return nil, false
+	}
+
+	// Note that we could check the current time here and discard the returned value
+	// if the expiration time has passed. But this would increase this function's execution
+	// time by > 50% (since time.Now is relatively expensive). Instead, we don't check time
+	// here and accept some imprecision in actual eviction times.
```

Comment:
i think this is related to another of my comments: If we had an `Entry` object that had eviction time in nanos on it, it might allow for comparison (where needed/wanted) at the usage time (maybe when the caller had a time already, like `request.time` or something).

Of course, that would add a layer of indirection for all `Get` operations.

Other option: `GetWithExpiry() (interface{}, int64, bool)` or maybe `StrictGet()`

Reply:
The problem is that the expiration time we track internally has been normalized because of the use of baseTime. Let's see if there's a use for this functionality and we can also add it.

Reply:
:+1: